rfcs/websockets_http2.md

```diff
@@ -0,0 +1,24 @@
+# RFC #67: Add WebSockets over HTTP/2 support to wptserve
+
+## Summary
+
+Enable wptserve to support WebSockets over an existing http2 connection.
+
+## Details
+
+Allow tests to open a WebSocket over an existing http2 connection.
+
+wptserver needs to implement the [Bootstrapping WebSocket with HTTP/2 protocol](https://tools.ietf.org/html/rfc8441).
+
+It means that it needs to be modified to:
+- Set ENABLE_CONNECT_PROTOCOL settings to 1
+- Detect Extended CONNECT Method and implement the WebSocket opening handshake.
+- pywebsocket operates on Apache mod_python Request objects. So we need to wrap H2Response and H2Request into a Request-like object.
+
+## Risks
+
+WebSocket tests who source **websockets/constants.js** have a wss variant. We will need to add a new variant named **http2** in order to use the http2 port when opening the WebSockets. WTP server does not support insecure HTTP/2 so we don't need to expose an insecure HTTP/2 WebSocket variant.
```

Comment:
There are also websocket tests at `websockets/Secure-*` using wss, but not using `websockets/constants.js`. Instead they use `websockets/websocket.sub.js`, which might need a similar change.

(Why this disparity between websocket tests? Some tests came from Opera, other tests came from Microsoft, and nobody has rewritten them to be consistent...)

Reply:
What would be the ideal way to fix it? There is a lot of duplication between Secure-* and non Secure- test. Maybe they can be rewritten to use variants. That would reduce the number of files, and make it easier to test over h2?

Also, should we have a separate discussion and/or rfc for those tests? I'm ready to help with the rewriting of those tests. But it can be decoupled from the task of enabling ws over h2 in wptserve.

Reply:
> I'm ready to help with the rewriting of those tests.

Awesome! Thank you!

Yeah, I think: Merge `constants.js` and `websocket.sub.js` into one. Make the test files directly under `websocket/` be more like the tests in directories (e.g. using `variant`s). And move them into the directories. Maybe all tests could use some massage in the process, but at a high level.

There might be some duplication in the websocket handlers, too.

I'm not sure if an RFC is required for this kind of cleanup, or if an issue in wpt/ is sufficient.

Reply:
No RFC required for test cleanup.

Reply:
```suggestion
WebSocket tests who source **websockets/constants.js** have a wss variant. We will need to add a new variant named **http2** in order to use the http2 port when opening the WebSockets. WPT server does not support insecure HTTP/2 so we don't need to expose an insecure HTTP/2 WebSocket variant.
```